src/sysupdate/sysupdate-transfer.c

```diff
@@ -784,30 +791,236 @@ static void compile_pattern_fields(
         memcpy(ret->sha256sum, i->metadata.sha256sum, sizeof(ret->sha256sum));
 }
 
+typedef struct CalloutContext {
+        const Transfer *transfer;
+        const Instance *instance;
+        TransferProgress callback;
+        PidRef pid;
+        const char *name;
+        void* userdata;
+} CalloutContext;
+
+
+static CalloutContext *callout_context_free(CalloutContext *ctx)
+{
+        if (!ctx)
+                return NULL;
+
+        /* We don't own any data but need to clean up the job pid */
+        pidref_done(&ctx->pid);
+
+        return mfree(ctx);
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(CalloutContext*, callout_context_free);
+
+static int callout_context_new(const Transfer *t, const Instance *i, TransferProgress cb, const char *name, void* userdata, CalloutContext **ret)
+{
```

Comment:
as above, please place `{` at the end of the preceeding line.

also, we'd typically line break such long parameter lists